### DISCUSSION TITLE
Arrays from entangle doesn't seem to have built-in array objects

### DISCUSSION THREAD

**Discussion by korematter** [alpinejs/alpine#2018] [1. Help] [answered]
Hi there,

Noob here and I'm learning so much on these discussions. I'm wondering if someone could explain the issue I'm running into with Alpine JS and Livewire and sharing array data via entangle.  It seems arrays coming from Livewire via entangle doesn't have the built-in array functions like array.filter for example.  Is it not an array object? 

Here is a simple test blade file initializing the combos array in Alpine and everything works perfectly, most notabily the combosSelected function recognizes the this.combos.filter built-in array function.

test.blade.php
`<div x-data="{

    combos: [
        {
            id: 1,
            levelId: 1,
            styleId: 1,
            danceId: 1,
            danceName: 'C',
            selected: false
        },
        {
            id: 2,
            levelId: 1,
            styleId: 1,
            danceId: 2,
            danceName: 'R',
            selected: false
        },
    ],

    get combosSelected() {
        return this.combos.filter(combo => combo.selected);
    }

}">

    <template x-for="combo in combos">
        <button
            type="button"
            x-text="combo.danceName"
            x-on:click="combo.selected = !combo.selected"
            :class="combo.selected && 'bg-red-200'"
            class="px-2 py-1 border rounded">
        </button>
    </template>

    <span x-text="combosSelected.length"></span>

</div>
`

But when I  initialize the array in Livewire  and pass the data to Alpine like so

Test.php Livewire component
`class Test extends Component
{
    public $combos;

    public function mount()
    {

        $combos[] = [
            'id' => 1,
            'levelId' => 1,
            'styleId' => 1,
            'dancedId' => 1,
            'danceName' => 'C',
            'selected' => false
        ];

        $combos[] = [
            'id' => 2,
            'levelId' => 1,
            'styleId' => 1,
            'dancedId' => 2,
            'danceName' => 'R',
            'selected' => false
        ];

        $this->combos = $combos;

        //dd($this->combos)
    }

    public function render()
    {
        return view('livewire.test');
    }
}`

`<div x-data="{

    combos: @entangle('combos').defer,

    get combosSelected() {
        return this.combos.filter(combo => combo.selected);
    }

}">

    <template x-for="combo in combos">
        <button
            type="button"
            x-text="combo.danceName"
            x-on:click="combo.selected = !combo.selected"
            :class="combo.selected && 'bg-red-200'"
            class="px-2 py-1 border rounded">
        </button>
    </template>

    <span x-text="combosSelected.length"></span>

</div>`

I get this error where it's not recognizing the this.combos.filter built-in array function:
![Screen Shot 2021-09-05 at 2 50 22 PM](https://user-images.githubusercontent.com/58536534/132138099-a77e60dc-5d3f-4537-bff6-968bed7202b3.png)

Are the arrays initialized in Alpine and Livewire not the same? Is entangle not converting php arrays to a proper json array object?

Thanks in advance!
Steve

**Comment by SimoTod** [comment, marked as answer]
Have you tried to console.log this.combos inside the getter before the error to see what it returns (if it's a proxy object, use Alpine.raw(this.combos) but it's probably already unwrapped in the getter)?

**Reply by korematter** [reply]
Hi SimoTod,

Thanks for the quick reply!  It looks like a proxy object. 

I updated the getter with the console.log and just tried to return the Alpine.raw(this.combos) like so:

get combosSelected() {
        console.log(this.combos);
        return Alpine.raw(this.combos);
        {{-- return this.combos.filter(combo => combo.selected); --}}
    }

I get the following console log and the error message on the return:
![Screen Shot 2021-09-05 at 4 09 06 PM](https://user-images.githubusercontent.com/58536534/132140132-6e204955-ea76-431d-820e-2cae29d900c7.png)

The first object looks like a proxy object but unsure what the second object is all about. 

If I go back to the Alpine initialization and console log that it shows two of the same proxy objects:
![Screen Shot 2021-09-05 at 4 14 51 PM](https://user-images.githubusercontent.com/58536534/132140293-7c8a7d87-b291-495a-b1b6-1e4c2e186380.png)

The messages are the same with and without the Alpine.raw(this.combo).  Trying to understand this proxy object but I'm still a bit newish. Thanks!

**Reply by SimoTod** [reply]
Yeah, You don't want to returm Alpine.raw, that would break stuff (as it does). Can you console.log(Alpine.raw(this.combos))? Is that the expected array?

**Reply by korematter** [reply]
Right. I did the console.log(Alpine.raw(this.combos)) and the second object is the expected array but the first object is not (  the one with {initialValue:Array(2), _x_intercepter: true, initialized: f} ). Not sure what that's all about.
![Screen Shot 2021-09-05 at 5 50 50 PM](https://user-images.githubusercontent.com/58536534/132142335-d26bd6f5-ada0-4c5b-9ba6-6548fb8ba097.png)

If I console.log(Alpine.raw(this.combos)) from the Alpine initialized array, I get two of the same array objects which Alpine seems to like so something is different between Alpine initialized array and what coming out of entangle:
![Screen Shot 2021-09-05 at 5 53 17 PM](https://user-images.githubusercontent.com/58536534/132142389-9789eeb5-471e-4132-be04-b016fc96c58c.png)


If I had to guess, Alpine is not liking either the first object {initialValue:Array(2), _x_intercepter: true.....} or the second object is not a Proxy object coming from entangle.  Could this be a Livewire issue? Is there a workaround?

**Reply by SimoTod** [reply]
So, the unwrapped value is what you would expect if it was just vanilla javascript. In your case the second time is right but the first time is some sort of magic object created by entangle for a very short time. Entangle is defined in the livewire codebase by the way so I'm not sure if some livewire users can explain better what it is.

A simple workaround for now would be 
`return Array.isArray(this.combos) ? [] : this.combos.filter(...)`

**Reply by korematter** [reply]
Thanks SimoTod. Spot on. The workaround works perfectly (just needed to switch the logic). Don't know why I didn't think of just ignoring the magic object to make it work. Yeah, curious to know more about what entangle is doing. I found there was an issue with entangle where calebporzio says something about "deep cloning and references never get crossed"  but it's over my head. Anyways, thank you very much for your explanation and help.

https://github.com/livewire/livewire/issues/1682#issuecomment-701471673

"For context: the issue is because when using @entangle, Alpine was actually storing an exact reference to Livewire's array, so when Alpine mutated the array, it triggered the watcher, but checked to see that the values were different before sending the update to Laravel. Because both values are the same, it just skipped the Livewire update. I'm now deep cloning @entangle values so Alpine and Livewire references never get crossed."
